lib/spack/spack/config.py

```diff
@@ -207,61 +208,345 @@ def __repr__(self):
+            value = value.get(key, default)
+        return value
+
+    def set(self, path, value, scope=None):
+        """Convenience function for setting single values in config files.
+
+        Accepts the path syntax described in ``get()``.
+        """
+        section, _, rest = path.partition(':')
+
+        if not rest:
+            self.update_config(section, value, scope=scope)
+        else:
+            section_data = self.get_config(section, scope=scope)
+
+            parts = rest.split(':')
+            data = section_data
+            while len(parts) > 1:
+                key = parts.pop(0)
+                data = data[key]
+            data[parts[0]] = value
+
+            self.update_config(section, section_data, scope=scope)
+
+    def __iter__(self):
+        """Iterate over scopes in this configuration."""
+        for scope in self.scopes.values():
+            yield scope
+
+    def print_section(self, section):
+        """Print a configuration to stdout."""
+        try:
+            data = syaml.syaml_dict()
+            data[section] = self.get_config(section)
+            syaml.dump(data, stream=sys.stdout, default_flow_style=False)
+        except (yaml.YAMLError, IOError):
+            raise ConfigError("Error reading configuration: %s" % section)
+
+
+@contextmanager
+def override(path, value):
+    """Simple way to override config settings within a context."""
+    config = spack.config.get_configuration()
+    overrides = InternalConfigScope('overrides')
+    config.push_scope(overrides)
+    config.set(path, value, scope='overrides')
+
+    yield config
+
+    scope = config.pop_scope()
+    assert scope is overrides
+
+
+def get_configuration():
+    """This constructs Spack's standard configuration scopes
+
+    This is a singleton; it constructs one instance associated with this
+    module and returns it. It is bundled inside a function so that
+    configuratoin can be initialized lazily.
+
+    Return:
+        Configuration: object for accessing spack configuration
+
+    """
+    global _configuration
+    if not _configuration:
+        _configuration = Configuration()
+
+        # first do the builtin, hardcoded defaults
+        defaults = InternalConfigScope('_builtin', config_defaults)
+        _configuration.push_scope(defaults)
+
+        # Each scope can have per-platfom overrides in subdirectories
+        platform = spack.architecture.platform().name
+
+        # add each scope and its platform-specific directory
+        for name, path in configuration_paths:
+            _configuration.push_scope(ConfigScope(name, path))
+
+            plat_name = '%s/%s' % (name, platform)
+            plat_path = os.path.join(path, platform)
+            _configuration.push_scope(ConfigScope(plat_name, plat_path))
+
+        # we make a special scope for spack commands so that they can
+        # override configuration options.
+        _configuration.push_scope(InternalConfigScope('command_line'))
```

Comment:
@tgamblin to clarify I mean to say that any PR which wants to introduce custom configuration scopes will likely have to change the way this is handled (at which point they will have to add a unit test for the alternative handling for command line scopes). Not necessarily a problem, but FYI  (e.g. this would have to be resolved for environment support)

Reply:
Thanks -- I think this is ok